api/proto/info/info.proto

```diff
@@ -840,6 +840,14 @@ message ZInfoStorage {
   bool storageLocation = 4; // Storage location for app disks, images etc.
 }
 
+// Per snapshot information
+message ZInfoSnapshot {
```

Comment:
My expectation was that ZinfoSnapshot contains id as well as timestamp the snapshot was taken. Does the id itself include timestamp ?

Reply:
We can the timestamp later. I don't see a need for it right now.
Or, if you think we need if from the very beginning, I can add it.

Reply:
We need it from beginning. That's what we mentioned in internal design discussions. Basically EVE sends snapshotid:timestamp. That is essential since users will know when the snapshot was taken on device, not when they expected to be taken on UI (eventual consistency). It also helps when multiple snapshots are enabled, we need not revisit this API again.

Reply:
Ok, no problem, I will add it. But I still believe it's extra, as we do not even give control to snapshot creation in the first stage (it's created implicitly).

Reply:
Yes I agree its created implicitly. But customers should know when their volumes were snapshotted. For example if device lost access to cloud for few hours or even days. The data snapshotted will be different than what customer expected when he clicked update on UI. Giving timestamp helps in this case. Also we need to have timestamps anyway once multiple snapshots are allowed.

Reply:
Done

Reply:
LGTM.

Reply:
@OhmSpectator where did you add text about this id being (or including) a timestamp?

Reply:
@eriknordmark, I just added the timestamp. ID has nothing with the timestamp, it's not necessary if we just have a dedicated field for that.

Reply:
Which infomation will uniquely identify the snapshot? Only the ID=UUID? Or can we (in the future) have multiple snapshots with the same UUID and different snapshots?
If the latter, then the UUID does no longer uniquely identify the snapshot.
On idea is to define the identifier of a snapshot as the triple <volume UUID, timestamp, creator>, where the creator is an enum which can be EVE, controller, and down the road LPS. Thatway  the EVE APIs and internal controller APIs can uniformly use that composit identifier, which will uniquely identify each snapshot.

Reply:
I would say ID = UUID for the snapshot. I don't see a reason to add a timestamp or a creator label. We don't need a version either, as versioning the snapshots is not planned in any scenario.
It's also worth noting that snapshot UUID != volume UUID. Moreover, the snapshot UUID exposed to the controller may identify not a single snapshot file that corresponds to a specific volume but, in some cases, a group of snapshot files stored on a device. It's necessary for the case of several volumes attached to the guest. In this case, the snapshot UUID identifies the set of snapshot files created for all the volumes of the VM. But as we do not need to expose these low-level details to the Controller, we can use a generic snapshot UUID here.

Reply:
I think capturing that comment in the protobuf file is a good step; really like explicitly stating the part about multiple volumes which I had missed.

But I still don't see how that will be sufficient in the possible future when EVE might be scheduled to take snapshots periodically or a local profile server can ask for a snapshot (both when the controller might not be reachable).

I'd like to understand how we will identify the snapshots (for reporting and also selecting one to delete) if we don't plan for that evolution. (And in case it isn't obvious, I find that how we name/identify objects in a distributed system is one of the most important up-front design decisions. I don't know how many cases I've seen expendient things like e.g., IP addresses being used as identifiers and as we all know there is only one device was has IP 192.168.1.1 ;-) Here at least we start with UUIDs, but we need to look a bit more into the future to make sure it will last.

Reply:
Yep, I get your point. As I mentioned in the other comments, I do not see snapshot objects existing without a VM object that it belongs to. So I see a requirement for the ID to be unique only in the scope of one VM.
Do I miss something?